cluster: persist storage node IDs

zekker6 opened the pull request:
### Describe Your Changes

Related issue: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5438

Added a persistent storage node ID which can be used to identify storage nodes when computing target to write data. This allows to avoid data re-distribution in case storage node address have been changed.
By default, vminsert will still use ID generated based on an address specified with `-storageNode` command-line flag.

Storage node ID is communicated as a part of vmselect and vminsert handshakes. This means that this is a breaking change to the protocol and versions before this change will not be able to perform the handshake with newer ones.

Implementation wise:

- Cluster native endpoints in [multi-level setup](https://docs.victoriametrics.com/cluster-victoriametrics/#multi-level-cluster-setup) calculate vminsert and vmselect node IDs as a hash of all underlying storage node IDs, so that it will also remain consistent in case of address changes.

- `vmselect` does not use persistent node ID apart from multi-level setup use-case

- storage node ID is persisted at `<storageDataPath>/metadata/node_id.bin` so it is included as a part of backups without additional modifications

### Checklist

The following checks are **mandatory**:

- [x] My change adheres [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/contributing/).

zekker6 committed:
lib/storage: add storage node id

Generate random node ID on start if it is missing or load from disk. Save to storage on storage shutdown.

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

zekker6 committed:
lib/storage: store node ID in metadata so that it is included in the backups

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

zekker6 committed:
app/cluster: communicate node IDs when performing a handshake

Send a node ID of vmstorage as a part of vmselect and vminsert handshakes.
Use vmstorage node ID as an identifier for consistent hashing at vminsert.

Cluster native endpoints calculate vminsert and vmselect node IDs as a hash of all underlying storage node IDs, so that it will also remain consistent in case of address changes.

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

zekker6 committed:
docs/changelog: add info about persisting vmstorage node ID

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

zekker6 committed:
app/vminsert: disable usage of persistent storage node ID by default

This is needed in order to avoid complete data re-sharding after the upgrade to a new version.

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

f41gh7 commented on lib/storage/storage.go:
```diff
@@ -244,6 +249,22 @@ func MustOpenStorage(path string, retention time.Duration, maxHourlySeries, maxD
+		s.nodeID = encoding.UnmarshalUint64(nodeID)
+	} else {
+		nodeID := rand.Uint64()
+		s.nodeID = nodeID
```
Looks like new id must be saved immedeatly after it was generated with s.mustSaveNodeID()

Otherwise, it could be lost due to server crash.

 What is a chance of collision for math/rand?

f41gh7 replied:
how to reset id if needed?

zekker6 replied:
> What is a chance of collision for math/rand? 

`math/rand` is pseudo-random which is initualized with random source since Go 1.20 release, so we shouldn't expect any collisions here.

> how to reset id if needed?

Just delete `metadata/node_id.bin` and restart the storage. That would trigger generation of node ID again.

f41gh7 replied:
In this case, `mustSaveNodeID` cannot be called at storage shutdown (at first init is enough). Otherwise it'll override value and ll make it impossible to reset id for kubernetes or similar envs.

zekker6 replied:
While it does have lower chance of collision I would rather think if we really need the cance to be lower than uint64 already provides?
With 100 vmstorage nodes chance of collision will be 2.220446049250313e-16 (based on approximation of [Birthday problem](https://en.wikipedia.org/wiki/Birthday_problem)). It seems to me like having UUID here (122 random bits for v4 UUID) would not be that critical considering that node ID is generated once and cluster size is rarely exceeding even hundreds of nodes.

f41gh7 commented on app/vminsert/netstorage/netstorage.go:
```diff
@@ -568,8 +611,17 @@ func initStorageNodes(addrs []string, hashSeed uint64) *storageNodesBucket {
+		nodeIDs = append(nodeIDs, nodeID)
 		sns = append(sns, sn)
 	}
+	nodesHash := newConsistentHash(nodeIDs, hashSeed)
```
For now, it looks like if all vmstorages were down at the moment of vminsert start, all metric flow will be sent to the first node at nodeIDs array, since all nodeis == 0.

f41gh7 replied:
It's a complicated problem. Yes, it must be dynamically updated to be consistent for all vminsert nodes.

 Most probably, storageNodes with id == 0 could be placed into routing with broken state.

shichanglin5 replied:
> I've added a fallback in [bb8b5db](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/bb8b5db0fefc54726a31c5386e61c4c04a6bfb5b), do you think it will also be useful to implement a way for consistent has rebuild in order to get the real node IDs once vmstorage nodes will be available?

@zekker6 
Will generating a nodeId as a fallback cause other issues? The generated nodeId is completely different from the actual nodeId of the vmstorage node. In a consistent hash, this corresponds to a completely different node, which can lead to different time series shard distributions.

Additionally, if a vmstorage node fails after running a period of time, vminsert might still know the nodeId of that vmstorage. If some vminsert nodes restart, will this cause inconsistency in the nodeId view among different vminsert nodes?

zekker6 replied:
@shichanglin5 We've already discussed this and came up with exactly the same issues as you've described. Becase of this, the behaviour will be changed(currently WIP) to the following:
- if using persistent storage node IDs is enabled and storage node is not available on start it will not be a part of nodes pool
- once node will become reachable it will be added in pool with its persistent ID

This allows to ensure that persistent IDs will always be in sync and will minimize rerouting due to changes in consistent hash.

shichanglin5 replied:
nice

zekker6 committed:
lib/storage: save storage ID after init

This helps to avoid re-creating a storage ID in case of unclean shutdown.

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

zekker6 committed:
lib/storage: fallback to address-based ID

Generate an ID based on storage node address if storage node is not available.
This is needed in order to prevent uneven load distribution if some storage nodes are not available when vminsert is starting.

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

zekker6 commented on docs/CHANGELOG.md:
```diff
@@ -31,12 +31,14 @@ See also [LTS releases](https://docs.victoriametrics.com/lts-releases/).
 ## tip
 
 **Update note 1: the `--vm-disable-progress-bar` command-line flag at `vmctl` was deprecated. Use `--disable-progress-bar` instead.**
+**Update note 2: release contains breaking change to inter-cluster communication. The `vmstorage` nodes with version lower than `tip` won't be able to communicate with `vmselect` and `vminsert` nodes with version lower than `tip`. See [this](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5438) issue for the details.**
```
>  But for users not using this feature, the upgrade temporarily disrupt ingestion. Is it a good idea to make higher version of vminsert to be compatible with the lower version of vmstorage(the persistent id is not enabled in both cases)? So users can upgrade vminsert first, and then vmstorage without breaking ingestion.

I don't see a way to check if node ID is supported in [current handshake implementation](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/d88147d8f932b6e9fc146ebd565c6552428fbe61/lib/handshake/handshake.go#L107-L137) as we don't have any kind of versioning of the handshake itself. Maybe it's a good time to add some kind of versioning of the handshake for future withing a single breaking change?
Wdyt @hagen1778 @f41gh7?

f41gh7 replied:
I think getStorageNodeID must be a part of RPC protocol and shouldn't included into handshake function.

f41gh7 commented:
Overall, I don't like the idea to make routing between vminsert and vmstorage uneven and unpredictable by relying and ID generated by vmstorage.  

 It brings a lot of bugs, possible re-routing storms, complicates code and solves very niche problem - service discovery with Nomad for vmstorage nodes.

It cannot work with vminsert and vmselect at proxy mode aka -clusterListenAddr flag.

 But vminsert/vmselect already have support for dynamic storage discovery.  So it's too late to say stop to it.

hagen1778 commented on app/vminsert/netstorage/netstorage.go:
```diff
@@ -398,7 +403,8 @@ func (sn *storageNode) dial() (*handshake.BufferedConn, error) {
 	}
-	bc, err := handshake.VMInsertClient(c, compressionLevel)
+	bc, id, err := handshake.VMInsertClient(c, compressionLevel)
+	sn.id.CompareAndSwap(0, id)
```
shall we always compare with 0? Or shall we compare with previously set id?

zekker6 replied:
The idea of comparing to `0` is to assign value only once.
With the latest version there is no fallback to another ID value, so the only path to assign and ID is to get it from the handshake.

hagen1778 commented on app/vminsert/netstorage/netstorage.go:
```diff
@@ -407,6 +413,23 @@ func (sn *storageNode) dial() (*handshake.BufferedConn, error) {
+		sn.checkHealth()
+	}
+
+	// If the id is still not populated after checkHealth than storage node is not reachable
```
```suggestion
	// If the id is still not populated after checkHealth then storage node is not reachable
```

hagen1778 commented on app/vminsert/netstorage/netstorage.go:
```diff
@@ -407,6 +413,23 @@ func (sn *storageNode) dial() (*handshake.BufferedConn, error) {
+	// If the id is still not populated after checkHealth than storage node is not reachable
+	// build a unique id based on the address
+	if sn.id.Load() == 0 {
+		id := xxhash.Sum64String(sn.dialer.Addr())
```
would it make sense to init IDs like this when creating `sn`?

zekker6 committed:
lib/storage: don't save persistent node ID on shutdown

It is supposed to be saved right after it was generated, there is no reason re-write it again.

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

zekker6 committed:
lib/storage: print node ID in startup log

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

zekker6 committed:
lib/handshake: use a json payload for metadata exchange

Update the handshake to use an arbitrary JSON payload to transfer metadata.
Handshake sends the metadata length first as an uint64 and then the metadata itself.

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

zekker6 committed:
app/vminsert/netstorage: print storage node IDs in logs

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

shichanglin5 commented:
This issue mainly deals with time series redistribution caused by non-static IP addresses. It seems to be non-existent when deploying on Kubernetes because nodes are specified using DNS names. For example:

```shell
-selectNode=vmselect-meta-0.vmselect-meta.vm-cluster-meta:8481,vmselect-meta-1.vmselect-meta.vm-cluster-meta:8481
-storageNode=vmstorage-meta-0.vmstorage-meta.vm-cluster-meta:8401,vmstorage-meta-1.vmstorage-meta.vm-cluster-meta:8401
```

If the service is not deployed using K8s StatefulSet and instead IP addresses are manually specified, is it possible to add some extra information to uniquely identify the nodes in the command line parameters? 

For example:

```shell
-selectNode=192.168.0.1:8481@node-1,192.168.0.2:8481@node-2
```

Then `vminsert` can parse the node names `node-1` and `node-2` based on the `@` delimiter. If no additional information is specified, `vminsert` can maintain the existing logic to ensure compatibility.

zekker6 commented:
Hi @shichanglin5, this is one of the options which was suggested in the [original issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5438) (you can find it by checking the original version before it was edited by clicking at "edited by valyala" and choosing the oldest version).

This idea was rejected in favor of automatically generated storage node IDs because of the following reasons:
- providing IDs manually can be error-prone and makes it harder to manage command-line flags values
- it does not work with dynamic storage nodes discovery mechanism(such as `dns+srv` discovery)

zekker6 committed:
app/vminsert/netstorage: exclude unavailable nodes from consistent hash on start

Exclude unhealthy storage nodes from consistent hash in case persistent storage node IDs are enabled.
This is needed in order to avoid uneven distribution of load due to default(uint64(0)) IDs assigned to storage nodes.

Remove generating fallback ID from node IP address as this will cause a re-distribution of series once storage node will become available and will change its ID.

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

zekker6 committed:
app/vminsert/netstorage: make linter happy

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

shichanglin5 commented on app/vminsert/netstorage/netstorage.go:
```diff
@@ -586,14 +632,65 @@ func initStorageNodes(addrs []string, hashSeed uint64) *storageNodesBucket {
+						sns:    oldSnb.sns,
+					}
+
+					newNodeIDs := append(snb.nodesHash.nodeHashes, sn.getID())
```
race condition

zekker6 committed:
app/vminsert/netstorage: use correct snb reference

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

zekker6 committed:
app/vminsert/netstorage: reinitialize snb on vmstorage connection restore

It is needed to rebuild snb in order to ensure that list of storage nodes and consistent hash are in sync.
Updating just consistent hash ring is not safe because it can cause misalignment of indexes of alive nodes in snb.sns and hash slots.

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

zekker6 committed:
app/vminsert/netstorage: sync comment

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

shichanglin5 commented on app/vminsert/netstorage/netstorage.go:
```diff
@@ -671,20 +671,14 @@ func initStorageNodes(addrs []string, hashSeed uint64) *storageNodesBucket {
-
-					if !storageNodes.CompareAndSwap(oldSnb, &snbNew) {
+					currentSnb := getStorageNodesBucket()
+					newSnb := initStorageNodes(addrs, hashSeed)
```
If the `sn` changes to ready, `initStorageNodes` will be executed. I noticed that `initStorageNodes` contains some code for metrics registration（`ms.NewGauge(fmt.Sprintf(`vm_rpc_rows_pending{name="vminsert", addr=%q}`, addr)`）. Will multiple registrations cause any issues (like panic)? Perhaps adding a new method `updateStorageNodes` would be better.

zekker6 replied:
Hello @shichanglin5 

> I noticed that initStorageNodes contains some code for metrics registration（'ms.NewGauge(fmt.Sprintf(vm_rpc_rows_pending{name="vminsert", addr=%q}, addr)'）. Will multiple registrations cause any issues (like panic)? 

No, this will not cause panic as long as metrics belong to different `MetricSet`s.
 
> Perhaps adding a new method `updateStorageNodes` would be better. 

It would be more efficient, but it makes code much more complicated. It will require to introduce synchronization for `snb.sns` and `snb.nodesHash`, which means that there will be performance penalty at the hot path. Also it will make the code more complicated as all functions that use `snb` will require to acquire a lock (or an atomic copy of `sns` and/or `nodesHash`).
Considering that the case when storage nodes are not available on vminsert start is quite rare (it is always recommended to start storage nodes first) it seems like it efforts of doming more efficient update just don't make it.

f41gh7 commented on app/vminsert/clusternative/request_handler.go:
```diff
@@ -25,7 +26,7 @@ var (
 	// There is no need in response compression, since
 	// lower-level vminsert sends only small packets to upper-level vminsert.
-	bc, err := handshake.VMInsertServer(c, 0)
+	bc, err := handshake.VMInsertServer(c, 0, netstorage.GetNodeID())
```
this hash is not persistent. Each update to underlying storage nodes breaks it.

So it makes no sense to return it.

zekker6 replied:
> this hash is not persistent. Each update to underlying storage nodes breaks it.

Yes, this hash is not persisted, but it is expected that underlying storage node IDs are persisted, so building a hash based on these IDs should return the same result. Could you elaborate on each update to storage nodes forcing to change the ID?

f41gh7 commented on app/vminsert/netstorage/netstorage.go:
```diff
@@ -586,14 +632,59 @@ func initStorageNodes(addrs []string, hashSeed uint64) *storageNodesBucket {
+		go func(sn *storageNode) {
+			defer wg.Done()
+			for {
+				sn.brLock.Lock()
```
this code uses multiple mutex locks (sn.brlock + sn.brCond) + atomic pointer loads.

 It opens wide possibility to data races, race conditions, dead locks. 

 It's better to rewrite it and add tests.

f41gh7 replied:
It's better to write it with:
1) move logic to the separate function
2) add comments if needed. 
3) add tests

f41gh7 commented on app/vmselect/netstorage/netstorage.go:
```diff
@@ -2946,6 +2950,31 @@ func getStorageNodes() []*storageNode {
+)
+
+// GetNodeID returns unique identifier for underlying storage nodes.
+func GetNodeID() uint64 {
```
it's better to return always 0 value for consistency. Multi level cluster setup is not intended to use with persistent storage node ids.

zekker6 committed:
app/vmselect: send static empty node ID for multi-level setup

Multi-level vmselect setup is not intended to use storage node IDs, so it is safe to return 0 here.

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

zekker6 committed:
app/vminsert/netstorage: refactor snb rebuild

Signed-off-by: Zakhar Bessarab <z.bessarab@victoriametrics.com>

hagen1778 commented:
This PR is currently on-hold due to the following reasons:
1. It complicates the code
2. It has edge cases with multi-level cluster setup
3. It is not applicable to k8s, where IP change doesn't affect the service address. And k8s is used by majority of users.

f41gh7 commented:
While this feature could be useful and requires zero configuration from user perspective. It makes logic implicit and hard to debug. It also adds non-trivial changes to the all cluster components.

 I'd like to use aliases feature proposed at https://github.com/VictoriaMetrics/VictoriaMetrics/pull/10914
